nimbus-consensus/src/manual_seal.rs

```diff
@@ -0,0 +1,94 @@
+use std::sync::Arc;
+use sp_keystore::SyncCryptoStorePtr;
+use sp_runtime::{
+	traits::{Block as BlockT, DigestFor},
+	generic::{Digest, DigestItem},
+};
+use sp_core::crypto::Public;
+use sc_consensus::BlockImportParams;
+use sc_consensus_manual_seal::{ConsensusDataProvider, Error};
+use sp_api::{TransactionFor, ProvideRuntimeApi, HeaderT};
+use sp_inherents::InherentData;
+use nimbus_primitives::{NimbusApi, NimbusId, CompatibleDigestItem, NIMBUS_ENGINE_ID};
+
+/// Provides nimbus-compatible pre-runtime digests for use with manual seal consensus
+pub struct NimbusManualSealConsensusDataProvider<C> {
+	/// Shared reference to keystore
+	pub keystore: SyncCryptoStorePtr,
+
+	/// Shared reference to the client
+	pub client: Arc<C>,
+
+	// Could have a skip_prediction field here if it becomes desireable
+}
+
+impl<B, C> ConsensusDataProvider<B> for NimbusManualSealConsensusDataProvider<C>
+where
+	B: BlockT,
+	C: ProvideRuntimeApi<B> + Send + Sync,
+	C::Api: NimbusApi<B, NimbusId>,
+	{
+	type Transaction = TransactionFor<C, B>;
+
+	fn create_digest(
+		&self,
+		parent: &B::Header,
+		_inherents: &InherentData,
+	) -> Result<DigestFor<B>, Error> {
+		// Fetch first eligible key from keystore
+		let maybe_key = crate::first_eligible_key::<B, C>(
+			self.client.clone(),
+			&*self.keystore,
+			parent,
+			// For now we author all blocks in slot zero, which is consistent with  how we are
+			// mocking the relay chain height which the runtime uses for slot beacon.
+			// This should improve. See https://github.com/PureStake/nimbus/issues/3
+			0,
```

Comment:
Actually, this isn't even good enough for now. It matches how we are mocking the relay block number here in this repo, but not in Moonbeam or other downstream projects. For the specific case of the relay block height, we might be able to extract the relay block height from the inherent.

Reply:
Idea: implement `ConsensusDataProvider` for each of our in-runtime slot beacons. They should each be able to extract the necessary data from either header or inherents.

SlotBeacon | How it works in the runtime | How it works in manual seal
--- | --- | ---
OwnBlockNumber | Uses frame system which uses header | reads from header
RelayBlockNumber | Uses parachain system, which uses parachain inherent | reads from parachain inehrent
TimeBased | Uses pallet timestamp which uses timestamp inherent | reads from timestamp inherent

So that means the way to proceed for now is to make this one parachain specific